### PR TITLE
Enabling overflow checks for debug

### LINKED ISSUE #3140: Turn on CheckForOverflowUnderflow on all projects in Debug
Right now, only System.Reflection.Metadata has CheckForOverflowUnderflow on in Debug.

I think there's value in doing this across the board. Expressions that are allowed to overflow by design should be explicitly marked `unchecked`.

cc @weshaggard @stephentoub

### PR DESCRIPTION
**Pull request by dennisdietrich**
Turns on overflow/underflow checks for all managed
projects via the main dir.props file. Also contains additional
unchecked statements.

Fix #3140

### CODE REVIEW

**Comment by dennisdietrich** [comment]
@dotnet-bot help

**Comment by dennisdietrich** [comment]
@dotnet-bot test outerloop CentOS7.1 Debug
@dotnet-bot test outerloop Debian8.4 Debug
@dotnet-bot test outerloop Fedora23 Debug
@dotnet-bot test outerloop Fedora24 Debug
@dotnet-bot test outerloop OpenSUSE13.2 Debug
@dotnet-bot test outerloop OpenSUSE42.1 Debug
@dotnet-bot test outerloop OSX Debug
@dotnet-bot test outerloop PortableLinux Debug
@dotnet-bot test outerloop RHEL7.2 Debug
@dotnet-bot test outerloop Ubuntu14.04 Debug
@dotnet-bot test outerloop Ubuntu16.04 Debug
@dotnet-bot test outerloop Ubuntu16.10 Debug
@dotnet-bot test outerloop Windows 7 Debug
@dotnet-bot test outerloop Windows_NT Debug
@dotnet-bot test outerloop Windows Nano 2016 Debug

**Comment by dennisdietrich** [comment]
@dotnet-bot test outerloop CentOS7.1 Debug
@dotnet-bot test outerloop Debian8.4 Debug
@dotnet-bot test outerloop Fedora23 Debug
@dotnet-bot test outerloop Fedora24 Debug
@dotnet-bot test outerloop OpenSUSE13.2 Debug
@dotnet-bot test outerloop OpenSUSE42.1 Debug
@dotnet-bot test outerloop OSX Debug
@dotnet-bot test outerloop PortableLinux Debug
@dotnet-bot test outerloop RHEL7.2 Debug
@dotnet-bot test outerloop Ubuntu14.04 Debug
@dotnet-bot test outerloop Ubuntu16.04 Debug
@dotnet-bot test outerloop Ubuntu16.10 Debug
@dotnet-bot test outerloop Windows 7 Debug
@dotnet-bot test outerloop Windows_NT Debug
@dotnet-bot test outerloop Windows Nano 2016 Debug

**Comment by dennisdietrich** [comment]
@dotnet-bot test outerloop CentOS7.1 Debug
@dotnet-bot test outerloop Debian8.4 Debug
@dotnet-bot test outerloop Fedora23 Debug
@dotnet-bot test outerloop Fedora24 Debug
@dotnet-bot test outerloop OpenSUSE13.2 Debug
@dotnet-bot test outerloop OpenSUSE42.1 Debug
@dotnet-bot test outerloop OSX Debug
@dotnet-bot test outerloop PortableLinux Debug
@dotnet-bot test outerloop RHEL7.2 Debug
@dotnet-bot test outerloop Ubuntu14.04 Debug
@dotnet-bot test outerloop Ubuntu16.04 Debug
@dotnet-bot test outerloop Ubuntu16.10 Debug
@dotnet-bot test outerloop Windows 7 Debug
@dotnet-bot test outerloop Windows_NT Debug
@dotnet-bot test outerloop Windows Nano 2016 Debug

**Comment by dennisdietrich** [comment]
Not seeing any more overflow exceptions, though that of course doesn't mean I've found them all (realistically, there are probably still some left). @tarekgh @karelz Can I please get a review of the code changes and build failures?

File: src/System.IO.FileSystem/tests/FileStream/LockUnlock.cs [hunk 1]
```diff
@@ -164,7 +164,7 @@ public void OverlappingRegionsFromOtherProcess_ThrowsException(long fileLength,
                         Assert.Throws<IOException>(() => fs2.Lock(long.Parse(secondPos), long.Parse(secondLen)));
                     }
                     return SuccessExitCode;
-                }, path, secondPosition.ToString(), secondLength.ToString()).Dispose();
+                }, $"\"{path}\"", secondPosition.ToString(), secondLength.ToString()).Dispose();
```
Comment (tarekgh) [inline src/System.IO.FileSystem/tests/FileStream/LockUnlock.cs]: >}, $"\"{path}\" [](start = 16, length = 15)

why you did that?

File: src/System.IO.FileSystem/tests/FileStream/LockUnlock.cs [hunk 2]
```diff
@@ -175,7 +175,7 @@ public void OverlappingRegionsFromOtherProcess_ThrowsException(long fileLength,
                         fs2.Unlock(long.Parse(secondPos), long.Parse(secondLen));
                     }
                     return SuccessExitCode;
-                }, path, secondPosition.ToString(), secondLength.ToString()).Dispose();
```
Comment (tarekgh) [inline src/System.IO.FileSystem/tests/FileStream/LockUnlock.cs]: >, path [](start = 17, length = 6)

ditto

**Comment by tarekgh** [comment]
Test OuterLoop Fedora23 Debug please
Test OuterLoop Fedora24 Debug please 
Test OuterLoop OpenSUSE13.2 Debug please
Test OuterLoop PortableLinux Debug please
Test OuterLoop Windows Nano 2016 Debug please

**Comment by tarekgh** [comment]
> Failure to restore a package: Unable to resolve 'coveralls.io (>= 1.4.0)' for '.NETStandard,Version=v2.0'. 

@mmitche is this known issue?

**Comment by danmosemsft** [comment]
yes https://github.com/dotnet/corefx/issues/16223

File: src/System.IO.FileSystem/tests/FileStream/LockUnlock.cs (hunk 1 above)
Comment (dennisdietrich) [inline src/System.IO.FileSystem/tests/FileStream/LockUnlock.cs]: Not putting the path in parenthesis is actually a bug and the test won't work in environments like mine where there is a space a in the path (see #15891).

File: src/System.IO.FileSystem/tests/FileStream/LockUnlock.cs (hunk 2 above)
Comment (dennisdietrich) [inline src/System.IO.FileSystem/tests/FileStream/LockUnlock.cs]: Also ditto. ;)

**Comment by tarekgh** [comment]
Test OuterLoop Fedora23 Debug please
Test OuterLoop Fedora24 Debug please
Test OuterLoop OpenSUSE13.2 Debug please
Test OuterLoop PortableLinux Debug please
Test OuterLoop Windows Nano 2016 Debug please

**Comment by tarekgh** [comment]
we still have issue with myget feeds so the failed tests are because of that. we'll wait fixing the issue and rerun these tests

**Comment by tarekgh** [comment]
Test OuterLoop Fedora23 Debug please
Test OuterLoop Fedora24 Debug please
Test OuterLoop OpenSUSE13.2 Debug please
Test OuterLoop PortableLinux Debug please
Test OuterLoop Windows Nano 2016 Debug please

**Comment by tarekgh** [comment]
test OuterLoop OpenSUSE13.2 Debug please

**Comment by tarekgh** [comment]
test OuterLoop Windows Nano 2016 Debug please

**Comment by tarekgh** [comment]
test OuterLoop OpenSUSE13.2 Debug please

**Comment by tarekgh** [comment]
The failure in OuterLoop PortableLinux Debug is the one tracked here 
https://github.com/dotnet/corefx/issues/16201

**Comment by tarekgh** [comment]
test OuterLoop OpenSUSE13.2 Debug please

**Comment by tarekgh** [comment]
OuterLoop Fedora24 Debug failure is the issue tracked by the issue https://github.com/dotnet/corefx/issues/16225

OuterLoop PortableLinux Debug failure is the one tracked by the issue https://github.com/dotnet/corefx/issues/16201

OuterLoop Fedora23 Debug failure is tracked by the issue https://github.com/dotnet/corefx/issues/13300

the other 2 failures are just restoring package failures.

**Pull request merged**